### PR TITLE
Build refactor

### PR DESCRIPTION
**Pull request by jjsjann123**
Allows nvfuser to build use setuptools. This is mostly targeting cases where user want to update nvfuser bits on a pytorch installation.

target workflow would be:
building nvfuser wheel:
```
pip install torch  # we'll need torch nightly after this PR: https://github.com/pytorch/pytorch/pull/97404
git clone https://github.com/NVIDIA/Fuser
cd fuser && python setup.py bdist_wheel  # this one generates ./dist/nvfuser-xxx.whl
```

using nvfuser wheel:
```
# assume this is an environment where pytorch is installed via a wheel package
pip install nvfuser-xxx.whl
patch-nvfuser
# use as-is
```
where,
 - nvfuser wheel installation distributes nvfuser python scripts to replace the corresponding part shipped by torch;
 - `patch-nvfuser` copies `libnvfuser_codegen.so` to replace codegen binary shipped by torch.

Note:
1. multidevice code needed `gloo` source, we have it included as a submodule and attached its LICENSE in our wheel package.
2. pybind11 is also needed for python API. PyTorch doesn't ship pybind11 code base. `pip install pybind11[global]` would setup pybind11 cmake config.

TODO:
- [x] clean up `./nvfuser_python_utils/` to move __init__.py out
- [x] setup benchmark & cpp test through top level cmake build
- [x] double check submodule license

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            "clean": clean,
+        },
+        package_data={
+            "nvfuser": nvfuser_package_data,
+        },
+        entry_points={
+            'console_scripts': [
+                'patch-nvfuser = nvfuser_python_utils:patch_pytorch_nvfuser_binaries',
+            ],
+        },
```
Comment (crcrpar) [inline setup.py:203]: it might be better to explicitly mention pybind11[global] as a dependency
```suggestion
        },
        install_requires=["pybind11[global]"],
```

QQ: why is `[global]` needed? 

ref: https://setuptools.pypa.io/en/latest/userguide/dependency_management.html

File: csrc/utils.h [hunk 2]
```diff
@@ -28,6 +30,13 @@ bool is_zero_sized_tensor(const std::shared_ptr<c10::TensorType>& tensor_type);
 bool is_cpu_scalar(const at::Tensor& tensor);
 bool is_cpu_scalar(const c10::TensorType& tensor_type);
 
+// query codegen output arch and target
```
Comment (naoyam) [inline csrc/utils.h:33]:
```suggestion
// Query the target GPU version number NVRTC compiles generated CUDA kernels for
```

Also, the function name doesn't seem to be representing well what it actually does. Maybe something like `queryTargetGPUVersion`?

Does this need to be here? Isn't it only used by executor.cpp?

File: csrc/utils.cpp [hunk 3]
```diff
@@ -197,6 +199,61 @@ const auto& getEnableOptions() {
...
+// query codegen output arch and target
+void codegenOutputQuery(
+    const cudaDeviceProp* const prop,
+    int& major,
+    int& minor,
+    bool& compile_to_sass) {
+#ifdef USE_ROCM
+  NVRTC_SAFE_CALL(nvrtcVersion(&major, &minor));
+  compile_to_sass = false;
+#else
```
Comment (naoyam) [inline csrc/utils.cpp:211]: Remove

File: csrc/macro.h
Comment (naoyam) [inline csrc/macro.h:1]: Do you need a new file for this? Why don't we just have them in utils.h? Or if they are only used by executor, executor_utils.h may be a better location?

File: csrc/macro.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  do {                                                            \
+    nvrtcResult _result = x;                                       \
+    if (_result != NVRTC_SUCCESS) {                                \
+      std::cerr << "\nerror: " #x " failed with error "           \
+                << nvrtcGetErrorString(_result) << '\n';           \
+      exit(1);                                                    \
+    }                                                             \
+  } while(0)
+
+#define CUDA_SAFE_CALL(x)                                         \
```
Comment (naoyam) [inline csrc/macro.h:11]: We have runtime and driver calls, so we should probably need this for the runtime API as well.

File: third_party/gloo
Comment (naoyam) [inline third_party/gloo:1]: Was this just accidentally added?

File: csrc/executor.cpp [hunk 5]
```diff
@@ -1467,7 +1465,7 @@ float FusionExecutor::runRtc(
 
   cudaEventRecord(start_event, stream);
 
-  AT_CUDA_DRIVER_CHECK(at::globalContext().getNVRTC().cuLaunchKernel(
+  cuLaunchKernel(
```
Comment (naoyam) [inline csrc/executor.cpp:1468]: Please use CUDA_SAFE_CALL

File: csrc/executor.cpp [hunk 6]
```diff
@@ -368,10 +367,10 @@ void FusionExecutor::compileFusion(
 
   // The driver API call requires an int argument.
   int max_dynamic_smem = 0;
-  AT_CUDA_DRIVER_CHECK(at::globalContext().getNVRTC().cuFuncGetAttribute(
+  cuFuncGetAttribute(
```
Comment (naoyam) [inline csrc/executor.cpp:370]: Please use CUDA_SAFE_CALL (and also all the other API calls)

File: csrc/executor_utils.cpp [hunk 7]
```diff
@@ -916,7 +917,7 @@ void initializeCudaContext() {
   // lazily construct context if non-existing yet;
   // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
   CUcontext pctx = nullptr;
-  AT_CUDA_DRIVER_CHECK(at::globalContext().getNVRTC().cuCtxGetCurrent(&pctx));
+  cuCtxGetCurrent(&pctx);
```
Comment (naoyam) [inline csrc/executor_utils.cpp:920]: This function (`initializeCudaContext`) doesn't seem to be used. There's calls to `at::cuda::jit::initializeCudaContext`, though.

File: csrc/utils.h (hunk 2 above)
Comment (jjsjann123) [reply csrc/utils.h:33]: That's a good point.

This file was shared among multiple backends previously (NNC and legacy fuser). But since that file is not shipped with binary, I need to move it here. I'll dump this inside `executor.cpp` then.

File: third_party/gloo
Comment (jjsjann123) [reply third_party/gloo:1]: Nope, we actually needed this.

> multidevice code needed gloo source, we have it included as a submodule and attached its LICENSE in our wheel package.

I believe it's coming from https://github.com/NVIDIA/Fuser/blob/main/csrc/multidevice/ProcessGroupBuilder.cpp#L30-L34
That macro is inherited from `torch_cpu`

File: setup.py (hunk 1 above)
Comment (jjsjann123) [reply setup.py:203]: I believe `pybind11[global]` gives us both `pybind11` and `pybind11-global` with pip.

`pybind11-global` is needed for cmake, since it contains `Pybinbd11Confit.cmake`. pytorch doesn't ship pybind11 with its wheel.

File: third_party/gloo
Comment (naoyam) [reply third_party/gloo:1]: @samnordmann Do we need to use Gloo? Looks like we just need some process group.

File: setup.py (hunk 1 above)
Comment (jjsjann123) [reply setup.py:203]: Can't really use setuptool for the requirement as suggested here. Since we need pybind11 for cmake, which happens before we call `setup`.
I added `requirements.txt`. so we should just run `pip install -r requirements.txt`.

File: third_party/gloo
Comment (zasdfgbnm) [inline third_party/gloo:1]: Are we using gloo somewhere?

File: .gitignore
Comment (crcrpar) [inline .gitignore:1]: https://github.com/github/gitignore/blob/main/Python.gitignore would be a nice reference

File: .gitmodules [hunk 8]
```diff
@@ -0,0 +1,9 @@
+[submodule "third_party/gloo"]
+	path = third_party/gloo
+	url = https://github.com/facebookincubator/gloo
```
Comment (crcrpar) [inline .gitmodules:3]: nit-picking/question
What will happen when
- fuser is not `PROJECT_IS_TOP_LEVEL`
- I run `USE_DISTRIBUTED=0 python setup.py install`

?

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        cmd_str = [
+            get_cmake_bin(),
+            "--build",
+            build_dir_name,
+            "--target",
+            "install",
+            "--",
+            "-j",
+            max_jobs,
+        ]
```
Comment (crcrpar) [inline setup.py:183]: optionally will be better if we can set `MAX_JOBS` via an environment variable like https://github.com/pytorch/pytorch/blob/96e3b3ac724c4e262dc8fc05da18ebcbda6473b8/setup.py#L11-L12

File: csrc/executor_utils.cpp [hunk 10]
```diff
@@ -138,6 +138,55 @@ std::string kernelPreamble() {
...
+TORCH_CUDA_CU_API void queryTargetGPUVersion(
+    const cudaDeviceProp* const prop,
+    int& major,
+    int& minor,
+    bool& compile_to_sass) {
+  using CudaVersion = std::pair<int, int>;
+  CudaVersion nvrtc_version;
+  NVRTC_SAFE_CALL(nvrtcVersion(&nvrtc_version.first, &nvrtc_version.second));
+
+  TORCH_CHECK(
```
Comment (crcrpar) [inline csrc/executor_utils.cpp:151]: Will `TORCH_*` macros be removed in the end?

**Review by crcrpar — COMMENTED** [review]
dropping my trivial questions and comments

File: CMakeLists.txt [hunk 11]
```diff
@@ -1,39 +1,64 @@
...
+  message(STATUS "top-level build")
+  find_package(Torch REQUIRED)
+  find_package(Python REQUIRED Development Interpreter)
+  find_package(pybind11 REQUIRED)
+  # need this since the pytorch execution uses a different name
+  set(PYTHON_EXECUTABLE ${Python_EXECUTABLE})
+  set(ATEN_CUDA_ROOT "${TORCH_INSTALL_PREFIX}/include/ATen")
+  if(BUILD_NVFUSER_BENCHMARK)
+    # need this to download googletest
+    set(BENCHMARK_DOWNLOAD_DEPENDENCIES ON)
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply CMakeLists.txt:26]: That's used by `third_party/benchmark`. I think it's trying to pull googletest.

File: .gitmodules (hunk 8 above)
Comment (jjsjann123) [reply .gitmodules:3]: That's a good question!

When fuser is not top level build, so we are basically building from Pytorch where nvfuser `CMakeLists.txt` is added.
In that case, we do NOT need any submodule inside nvfuser and we'll just use gloo provided by pytorch.

When we switch the upstream pytorch `third_party/nvfuser` to a submodule, I don't know if there's any way for us to stop a totally unnecessary recursive submodule update.

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        subprocess.check_call(cmd_str)
+
+        # copy nvfuser pybind extension
+        src = os.path.join(cwd, "nvfuser", "lib", "libnvfuser.so")
+        dst = os.path.join(cwd, "nvfuser", "_C.cpython-310-x86_64-linux-gnu.so")
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply setup.py:188]: Hahaha, you caught me there. I'll update it.

File: .gitmodules (hunk 8 above)
Comment (crcrpar) [reply .gitmodules:3]: I'm a bit concerned with the possibility that we could have the divergent Gloo submodules whose side effects are unclear to me (partly because of my lack of knowledge)

ah, you've added the TODO for this: https://github.com/NVIDIA/Fuser/pull/59/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR192-R193

Comment (jjsjann123) [reply .gitmodules:3]: > Gloo submodules whose side effects are unclear to me (partly because of my lack of knowledge)

I don't think I know better than you. And there is a possibility that any library we link against might differ from what upstream is using. Frankly speaking I don't have a good solution, other than sync'ing our submodule commit with upstream pytorch. :cry:

File: test/main.cpp [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  } else {
+    filter.push_back(':');
+  }
+  filter += flag;
+  return filter;
+}
+
+int main(int argc, char* argv[]) {
+  ::testing::InitGoogleTest(&argc, argv);
+  if (!torch::cuda::is_available()) {
```
Comment (naoyam) [inline test/main.cpp:21]: Can we just make this an error? I don't see any use case where we want to potentially run with no CUDA.

File: csrc/utils.h [hunk 14]
```diff
@@ -18,6 +18,8 @@
 #include <typeinfo>
 #include <vector>
 
+#include <cuda_runtime.h>
```
Comment (naoyam) [inline csrc/utils.h:21]: Why is this necessary?

File: csrc/utils.cpp [hunk 15]
```diff
@@ -9,6 +9,7 @@
 #include <utils.h>
 
 #include <c10/util/string_view.h>
+#include <nvrtc.h>
```
Comment (naoyam) [inline csrc/utils.cpp:12]: Why is this necessary?

File: csrc/executor_utils.h [hunk 16]
```diff
@@ -27,6 +27,39 @@
...
+      exit(1);                                           \
+    }                                                    \
+  } while (0)
+
+#define CUDA_SAFE_CALL(x)                                               \
+  do {                                                                  \
+    CUresult _result = x;                                               \
+    if (_result != CUDA_SUCCESS) {                                      \
+      const char* msg;                                                  \
+      cuGetErrorName(_result, &msg);                                    \
```
Comment (naoyam) [inline csrc/executor_utils.h:46]: Did you intend to use `cuGetErrorString`? Maybe use both `cuGetErrorName` and `cuGetErrorString` as in the CUDA RT version?

**Review by crcrpar — APPROVED** [review]
looks good to me. I think it's good to go once we verify https://github.com/NVIDIA/Fuser/wiki/Building-fuser-project#option-2---build-nvfuser-separately-from-pytorch-pending-pr-59.

File: csrc/executor_utils.cpp [hunk 17]
```diff
@@ -1240,7 +1271,7 @@ std::tuple<NvrtcFunction, std::string, std::vector<char>> nvrtcCompile(
         }
         std::cout << log.data() << std::endl;
       }
-      AT_CUDA_NVRTC_CHECK(result);
+      NVRTC_SAFE_CALL(result);
```
Comment (naoyam) [inline csrc/executor_utils.cpp:1274]: Is there any specific reason to do this check here? There's also a check at line 1262.

File: csrc/executor_utils.cpp [hunk 18]
```diff
@@ -1281,8 +1312,7 @@ std::tuple<NvrtcFunction, std::string, std::vector<char>> nvrtcCompile(
     }
 
     const char* lowered_kernel_name = nullptr;
-    at::globalContext().getNVRTC().nvrtcGetLoweredName(
-        program, func_name.c_str(), &lowered_kernel_name);
+    nvrtcGetLoweredName(program, func_name.c_str(), &lowered_kernel_name);
```
Comment (naoyam) [inline csrc/executor_utils.cpp:1315]: Don't we want to use `NVRTC_SAFE_CALL`?

File: csrc/executor_utils.cpp [hunk 19]
```diff
@@ -1212,21 +1244,20 @@ std::tuple<NvrtcFunction, std::string, std::vector<char>> nvrtcCompile(
...
         std::string name = ss.str();
         FUSER_PERF_SCOPE("executor_utils::NvrtcCreateProgram");
-        AT_CUDA_NVRTC_CHECK(at::globalContext().getNVRTC().nvrtcCreateProgram(
+        NVRTC_SAFE_CALL(nvrtcCreateProgram(
             &program, code.c_str(), name.c_str(), 0, nullptr, nullptr));
       }
 
-      at::globalContext().getNVRTC().nvrtcAddNameExpression(
-          program, func_name.c_str());
+      nvrtcAddNameExpression(program, func_name.c_str());
```
Comment (naoyam) [inline csrc/executor_utils.cpp:1251]: Don't we want to use `NVRTC_SAFE_CALL`?

File: csrc/executor_utils.cpp (hunk 18 above)
Comment (jjsjann123) [reply csrc/executor_utils.cpp:1315]: Good catch~~
I don't think we had nvrtc checks before, and I was only doing a string replacement. Let me revisit those

File: csrc/executor_utils.cpp (hunk 17 above)
Comment (jjsjann123) [reply csrc/executor_utils.cpp:1274]: No. I think we should have replaced 1262 with this one.

File: csrc/executor_utils.h (hunk 16 above)
Comment (jjsjann123) [reply csrc/executor_utils.h:46]: Good point, will add that.

File: csrc/utils.cpp (hunk 15 above)
Comment (jjsjann123) [reply csrc/utils.cpp:12]: Because we are calling nvrtc APIs directly now, instead of going through the pytorch stub `at::globalContext().getNVRTC()`, so we'll need to have nvrtc header directly.

Comment (naoyam) [reply csrc/utils.cpp:12]: Yes, but does utils.cpp need it? I looked at the file but couldn't see where it might be necessary.

Comment (jjsjann123) [reply csrc/utils.cpp:12]: :exploding_head: ha, good catch... Let me double check the two headers added by me.

File: csrc/executor_utils.cpp (hunk 17 above)
Comment (naoyam) [reply csrc/executor_utils.cpp:1274]: We should probably just this line. Line 1262 doesn't just assert on the return value, but it also gives the "log" of the compilation.

**Review by naoyam — APPROVED** [review]
LGTM. Does this mean the current build instruction no longer works?

**Comment by jjsjann123** [comment]
> LGTM. Does this mean the current build instruction no longer works?

Current build instruction should still work. I'm double checking that right now.

On top of that, we can also build nvfuser against a installed pytorch. (current pip wheel from pytorch nightly might not work yet, there's some issue with their packaging. I'm patching that here: https://github.com/pytorch/pytorch/pull/97404, which is already merged. So the moment of truth would be tomorrow)

**Comment by jjsjann123** [comment]
Local tests passed. I'm merging this one.

I'll double check the pip wheel thing works with upstream pytorch. :crossed_fingers:

File: third_party/gloo
Comment (samnordmann) [reply third_party/gloo:1]: Sorry for the late answer! I wasnt receiving notifications somehow...
In my opinion we shouldn't rely on Gloo directly; we should stick to pytorch's ProcessGroup (that in turn can choose Gloo, NCCL or UCC as a backend)
Does it answer the question ?

Comment (crcrpar) [reply third_party/gloo:1]: I guess so though https://pytorch.org/docs/stable/distributed.html#torch.distributed.is_gloo_available might work better for this case?

Comment (samnordmann) [reply third_party/gloo:1]: > I think the question is what the minimum requirements for the multi-device runtime. Our build would become separated from PyTorch itself, and I think we need the separately built PyTorch binary must be compiled with, e.g., `USE_DISTRIBUTED=1`. Is there anything else?

I think that's it, but effectively you also need at least one backend. For now, I only have tested NCCL backend, Gloo should work as is, but UCC may need a small adjustment

Comment (naoyam) [reply third_party/gloo:1]: OK, so we should remove the direct dependency to Gloo. Instead, we need to make sure the PyTorch binary is built with the required dependency. CC: @jjsjann123
